internal/characters/alhaitham/cons.go

```diff
@@ -0,0 +1,110 @@
+	"github.com/genshinsim/gcsim/pkg/core/attributes"
+	"github.com/genshinsim/gcsim/pkg/core/combat"
+	"github.com/genshinsim/gcsim/pkg/core/glog"
+	"github.com/genshinsim/gcsim/pkg/core/player/character"
+	"github.com/genshinsim/gcsim/pkg/modifier"
+)
+
+// When a Projection Attack hits an opponent, Universality: An Elaboration on Form's CD is decreased by 1.2s.
+// This effect can be triggered once every 1s.
+func (c *char) c1(a combat.AttackCB) {
+	//ignore if c1 on icd
+	if c.c1ICD > c.Core.F {
+		return
+	}
+	c.ReduceActionCooldown(action.ActionSkill, 72) //reduced by 1.2s
+	c.c1ICD = c.Core.F + 60                        //1s icd
+}
+
+// When Alhaitham generates a Chisel-Light Mirror, his Elemental Mastery will be increased by 50 for 8 seconds,
+// max 4 stacks.
+// Each stack's duration is counted independently.
+// This effect can be triggered even when the maximum number of Chisel-Light Mirrors has been reached.
+func (c *char) c2() {
+	m := make([]float64, attributes.EndStatType)
+	m[attributes.EM] = 50
+	c.AddStatMod(character.StatMod{
+		Base:         modifier.NewBaseWithHitlag(fmt.Sprintf("alhaitham-c2-%v-stack", c.c2counter+1), 480), //8s
+		AffectedStat: attributes.EM,
+		Amount: func() ([]float64, bool) {
+
+			return m, true
+		},
+	})
+	c.c2counter = (c.c2counter + 1) % 4 //stacks are independent from each other, this will cycle them
+}
+
+// When Particular Field: Fetters of Phenomena is unleashed, the following effects will become active
+// based on the number of Chisel-Light Mirrors consumed and created this time around:
+// ·Each Mirror consumed will increase the Elemental Mastery of all other nearby party members by 30 for 15s.
+// ·Each Mirror generated will grant Alhaitham a 10% Dendro DMG Bonus for 15s.
+// The pre-existing duration of the aforementioned effects will be cleared if you use Particular Field: Fetters of Phenomena again while they are in effect
+func (c *char) c4(src string, counter int) {
+	//TODO: Does C4 apply the dendro dmg bonus to alhaitham even if he is NOT on field during the refund?
+	m := make([]float64, attributes.EndStatType)
+	if src == "gain" {
+		m[attributes.DendroP] = 0.1
+		c.AddStatMod(character.StatMod{
+			Base:         modifier.NewBaseWithHitlag(fmt.Sprintf("alhaitham-c4-%v-instance", counter), 900), //15s
+			AffectedStat: attributes.DendroP,
+			Amount: func() ([]float64, bool) {
+
+				return m, true
+			},
+		})
+		c.Core.Log.NewEvent("c4 buff on gain", glog.LogCharacterEvent, c.Index)
+	} else {
+		m[attributes.EM] = 30
+		for i, char := range c.Core.Player.Chars() {
+			// skip Alhaitham
+			if i == c.Index {
+				continue
+			}
+			char.AddStatMod(character.StatMod{
+				Base:         modifier.NewBaseWithHitlag(fmt.Sprintf("alhaitham-c4-%v-instance", counter+1), 900),
+				AffectedStat: attributes.EM,
+				Amount: func() ([]float64, bool) {
+					return m, true
+				},
+			})
+		}
+		c.Core.Log.NewEvent("c4 buff on loss", glog.LogCharacterEvent, c.Index)
+	}
+
+}
```

Comment:
Please look at the mirrorGain/mirrorLoss refactor comment in skill.go first. Please consider passing the amount of mirrors that were generated and consumed to this function and splitting this function into two, because the consumption and generation of mirrors on burst happens at different points in time. Also, the stacks are not counted separately. It could look like this:
```
func (c *char c4Loss(consumed int) {
        if consumed <= 0 {
                return
        }
        m := make([]float64, attributes.EndStatType)
        m[attributes.EM] = 30 * consumed
        for i, char := range c.Core.Player.Chars() {
                // skip Alhaitham
                if i == c.Index {
                        continue
                }
                char.AddStatMod(character.StatMod{
                        Base:         modifier.NewBaseWithHitlag("alhaitham-c4-loss", 900),
                        AffectedStat: attributes.EM,
                        Amount: func() ([]float64, bool) {
                                return m, true
                        },
                })
        }
}

func (c *char) c4Gain(generated int) {
        if generated <= 0 {
                return
        }
        m := make([]float64, attributes.EndStatType)
        m[attributes.DendroP] = 0.1 * generated
        c.AddStatMod(character.StatMod{
                Base:         modifier.NewBaseWithHitlag("alhaitham-c4-gain", 900),
                AffectedStat: attributes.DendroP,
                Amount: func() ([]float64, bool) {
                        return m, true
                },
        })
}
```